How-to-find-the-commit-which-introduced-a-bug.md

```diff
@@ -0,0 +1,186 @@
+
+
+1. Based on console errors or analysis of the code, find out which line(s) of code are causing problems.
+2. Run `git blame` for that specific file and see if the line was recently introduced to the file.
+3. Once the commit is obtained from the “blame” view, checkout that commit
+    ```bash
+    git checkout <commit-id>
+    ```
+    and see if the error is present. If yes, then checkout its parent commit (one commit earlier – you can find out what this is with `git log`).
+
+4. If the parent commit doesn’t have errors, then you’ve confirmed that the error is due to that commit through which the lines are added to the codebase. Congratulations! Mention this info in the issue thread and include a link to the problematic commit.
+
+## Using GitHub Website
+    
+
+Using the GitHub website, you can view the blame information for a file directly in the GitHub interface. Let’s try and use this feature on a file (contact-page.component.html) in the Oppia Repository.
+
+If you want a **guided Practical** go [here](https://app.tango.us/app/workflow/Using-Blame-with-github-Website--f48fa87dc2e749e2af2dbd10281f4ec2), and then click on "**Guide me**".
+
+
+Let’s start: 
+
+
+1. **Open the GitHub repository in your web browser.**
+
+![Screenshot of the home page](images/findCommitWhichIntroducedBug/image2.png)
+
+
+2. **Navigate to the file(In this example we need to navigate to the core/templates/pages/contact-page/contact-page.component.html)**
+
+
+![Screenshot of the "core" folder](images/findCommitWhichIntroducedBug/image6.png)
+
+
+![Screenshot of the "templates" folder](images/findCommitWhichIntroducedBug/image3.png)
+
+
+![Screenshot of the "pages" folder](images/findCommitWhichIntroducedBug/image1.png)
+
+3. **Open the file by clicking on its name.**
+
+
+![Screenshot of the "contact-page-component" file](images/findCommitWhichIntroducedBug/image5.png)
+
+
+4. **Now, you're viewing the content of the file. To see the blame information:**
+
+    **In the upper right corner of the file view, find the Blame button. It's next to the History and Raw buttons**.
+
+
+![Screenshot of the "blame" option highlighted](images/findCommitWhichIntroducedBug/image7.png)
+
+
+**This will display the blame information for each line directly on the GitHub website. You can see the commit hash, author, and date for each line in the file.**
+
+
+![Screenshot of the commit log highlighted](images/findCommitWhichIntroducedBug/image8.png)
+
+
+**If you click on the "Blame Info" as shown in the above image you will be forwarded to the commit page where it belongs.**
+
+
+![Screenshot asking you to click on "blame info"](images/findCommitWhichIntroducedBug/image4.png)
+
+
+## Git Bisect
+
+**`git bisect` is a powerful Git command that helps you find the commit that introduced a bug or regression in your codebase. It uses a binary search algorithm to efficiently narrow down the range of commits where the issue was introduced.**
```

Comment:
Maybe also explain: This is more useful if you don't know which line of code is causing the error (otherwise, we recommend using the "git blame" approach above instead).

Reply:
Done